layerscape: support sysupgrade

yangbolu1991 opened the pull request:
This patch-set is to support sysupgrade for both squashfs rootfs on flash, and ext4 rootfs on SD card.
Please help to review and merge.
Thanks a lot!

yangbolu1991 committed:
layerscape: convert to squashfs rootfs for QSPI NOR boot

There had been an issue in Layerscape QSPI driver for very long
time, which made squashfs,jffs2 rootfs not work on QSPI NOR.
And the ubifs had been used as a workaround.
Now the issue has been fixed. So convert to use squashfs,jffs2
rootfs on QSPI NOR for Layerscape boards (LS1012ARDB/LS1046ARDB/
LS1088ARDB), and update u-boot bootargs for booting.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: add compatible for board ls1043ardb

Backport a dts patch which added compatible for ls1043ardb.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: support sysupgrade for squashfs rootfs

Support sysupgrade for all Layerscape boards with squashfs
rootfs.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: fix board name for SD card boot

Many Layerscape boards support both flash boot and SD card boot.
And different firmware and sysupgrade.bin are built for the two
boot methods. To identify them, a fix could be done on board name
by adding a postfix "-sdboot" for SD card boot.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
layerscape: support sysupgrade for SD card ext4 rootfs

Support sysupgrade for SD card ext4 rootfs.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
I added more patches to support sysupgrade, and updated pull request title and description.
Thanks a lot:)

yangbolu1991 commented:
Any comments?
Thanks!

yangbolu1991 commented:
Hi @ynezz  
Will you help to review the pull request? There has been no feedback for a long time since created.
Thanks a lot!

adrianschmutzler commented:
> Backport a dts patch which added compatible for ls1043ardb.

If that's a backport from the kernel, it might help to include the kernel version where it was added in the _patch_ file name, e.g. `724-v5.5-net-sfp-rename-T_PROBE_WAIT-to-T_SERIAL.patch`

adrianschmutzler commented on target/linux/layerscape/image/armv8_64b.mk:
```diff
@@ -1,17 +1,19 @@
   KERNEL_LOADADDR := 0x80080000
   KERNEL_ENTRY_POINT := 0x80080000
+  IMAGE/sysupgrade.bin := append-kernel | pad-to 16M | \
+    append-rootfs | pad-rootfs | check-size 50331649 | append-metadata
```
Where is the argument for check-size coming from?
Consider defining IMAGE_SIZE instead.

yangbolu1991 committed:
layerscape: define IMAGE_SIZE and LS_SYSUPGRADE_IMAGE_SIZE

Define 64m IMAGE_SIZE for flash firmware.bin since the flash size
is 64MB. Define 48m LS_SYSUPGRADE_IMAGE_SIZE for flash sysupgrade.bin
which contains maximum 16MB kernel and 32MB rootfs according to
memory map.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

ynezz commented:
@tmn505 @mcbridematt can you please check/test this?

yangbolu1991 commented:
Here are two logs for flash sysupgrade, and SD sysupgarde on ls1088ardb, I verified that time.
Just for your reference. I should have attached them.
Thanks.

[openwrt-202005-ls1088ardb-sysupgrade-ext4.txt](https://github.com/openwrt/openwrt/files/4888481/openwrt-202005-ls1088ardb-sysupgrade-ext4.txt)
[openwrt-202005-ls1088ardb-sysupgrade-squashfs.txt](https://github.com/openwrt/openwrt/files/4888482/openwrt-202005-ls1088ardb-sysupgrade-squashfs.txt)

tmn505 commented:
It looks ok. The only downside with this is, that once written dtb will remain the same after every sysupgrade. To remedy this, I would suggest using FIT image.

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git

adrianschmutzler commented on target/linux/layerscape/image/armv8_64b.mk:
```diff
@@ -86,6 +86,12 @@ define Device/ls1012afrwy
+    ls-append-sdhead $(1) | pad-to 16M | \
+    append-kernel | pad-to $(LS_SD_ROOTFSPART_OFFSET)M | \
+    append-rootfs | check-size $(LS_SD_IMAGE_SIZE) | append-metadata
+  SUPPORTED_DEVICES := fsl,ls1012a-frwy-sdboot
```
I don't find these -sdboot suffixed values anywhere else, particularly not in any compatible as you partially reuse the DTS files anywhere. Is this actually working for sysupgrade then?

yangbolu1991 replied:
In OpenWrt, the board name info is created in by /lib/preinit/02_sysinfo.
You can find below patch of this pull request to fix up the board name for SD card boot.
layerscape: fix board name for SD card boot

adrianschmutzler replied:
Okay, this was the missing link I was looking. While I don't really like that approach conceptually, it appears to do the job.

Can you explain to me how the filtering works there? How do you make sure -sdboot is only added for sdboot variants, and not for all devices.

(Despite, I'd have used a higher number for that script, as sorting between 02_sysinfo and 02_sysinfo_fixup is not really obvious).
